aten/src/ATen/native/NativeFunctions.h

```diff
@@ -0,0 +1,126 @@
+#pragma once
+
+#include "ATen/ATen.h"
+#include <vector>
+
+/**
+ * ATen native functions are ways to write ATen methods which make only
+ * make use of other ATen operations (e.g., it is not necessary to
+ * bind into TH/THC code or drop into CUDA kernels.)  These functions
+ * are written as both functions as well as cwrap fragments, which are
+ * then folded into the ATen code generation process; define a function
+ * here, and it will show up as a method on at::Tensor.
+ *
+ * At the moment, only type_method_definition_level: base is supported.
+ */
+
+namespace at {
+namespace native {
+
+// [NativeFunction]
+Tensor type_as(const Tensor& self, const Tensor& other);
+
+// [NativeFunction]
+Tensor expand_as(const Tensor& self, const Tensor& other);
+
+// [NativeFunction]
+std::vector<Tensor> split(const Tensor& self, int64_t split_size, int64_t dim=0);
+
+// [NativeFunction]
+std::vector<Tensor> chunk(const Tensor& self, int64_t chunks, int64_t dim=0);
+
+// [NativeFunction]
+int64_t size(const Tensor& self, int64_t dim);
+
+
+// [NativeFunction]
+int64_t stride(const Tensor& self, int64_t dim);
+
+
+// [NativeFunction]
+bool is_same_size(const Tensor& self, const Tensor& other);
+
+// [NativeFunction]
+Tensor permute(const Tensor& self, IntList dims);
+
+
+// [NativeFunction]
+Tensor expand(const Tensor& self, IntList size);
+
+// [NativeFunction]
+Tensor squeeze(const Tensor& self);
+
+// [NativeFunction]
+Tensor squeeze(const Tensor& self, int64_t dim);
+
+// [NativeFunction]
+Tensor & squeeze_(Tensor& self);
+
+
+// [NativeFunction]
+Tensor & squeeze_(Tensor& self, int64_t dim);
+
+// [NativeFunction]
+Tensor unsqueeze(const Tensor& self, int64_t dim);
+
+// [NativeFunction]
+Tensor & unsqueeze_(Tensor& self, int64_t dim);
+
+/*
+[NativeFunction]
+variants: function
+*/
+Tensor stack(TensorList tensors, int64_t dim=0);
+
+/*
+[NativeFunction]
+name: RoiPooling2d_forward
+variants: function
+type_method_definition_dispatch: {
+  - CPU: at::native::RoiPooling2d_forward_cpu
+  - CUDA: at::native::RoiPooling2d_forward_cuda
+}
+*/
+std::tuple<Tensor, Tensor> RoiPooling2d_forward_cpu(
+  const Tensor& input,
+  const Tensor& rois,
+  int64_t pooledHeight,
+  int64_t pooledWidth,
+  double spatialScale);
+
+std::tuple<Tensor, Tensor> RoiPooling2d_forward_cuda(
+  const Tensor& input,
+  const Tensor& rois,
+  int64_t pooledHeight,
+  int64_t pooledWidth,
+  double spatialScale);
+
+/*
+[NativeFunction]
+name: RoiPooling2d_backward
+variants: function
+type_method_definition_dispatch: {
+  - CPU: at::native::RoiPooling2d_backward_cpu
+  - CUDA: at::native::RoiPooling2d_backward_cuda
+}
+*/
+Tensor RoiPooling2d_backward_cpu(
```

Comment:
Something that I have wondered about while working on CuDNN port to ATen is how native functions should handle the foo_out convention, e.g., there is already a result tensor preallocated and the backward should writre into it. I see in this case you have simply opted to not support this at all. Are we ever planning to support this mode of use, or not? (No code changes necessary.)

Reply:
I think we can support it if we need it.  I have some branch that implements 95% of it, but the annoying part is the user has to write the _out version, which isn't as natural as writing the standard version.  So I've punted until we have a real use case.

Reply:
OK. I ask because the old APIs to CuDNN were implemented in the _out style. But looking at the call sites, this isn't actually used in any sort of efficient way, so I think probably the right thing to do is to return tensors instead and keep punting this.

Reply:
Also note that I did write an "out" version initially to mimic the THNN style, but @colesbury recommended against doing this.